src/Database/SqlserverCompiler.php

```diff
@@ -78,6 +77,34 @@ protected function _buildInsertPart($parts, $query, $generator)
     }
 
     /**
+     * Generates the DELETE part of a SQL query
+     *
+     * @param array $parts The parts to build
+     * @param \Cake\Database\Query $query The query that is being compiled
+     * @param \Cake\Database\ValueBinder $generator the placeholder generator to be used in expressions
+     * @return string
+     */
+    protected function _buildDeletePart($parts, $query, $generator)
+    {
+        $parts = $this->_stringifyExpressions((array)$parts, $generator);
+
+        if ($query->clause('limit') === null) {
+            return sprintf('DELETE %s', implode(', ', $parts));
+        }
+
+        $template = "WITH CTE AS (SELECT TOP %d * %s %s %s) DELETE %s FROM CTE %s";
+
+        return sprintf(
+            $template,
+            $parts[5],
+            $parts[2],
+            $parts[3],
+            $parts[4],
+            $parts[6]
```

Comment:
I was thinking the exact same thing. It's done like this for all the compilers, but it would be better if `parts` was an object with get/set. I just see that as a different PR.